samples/simple-chat/csharp/FunctionApp/Functions.cs

```diff
@@ -3,184 +3,89 @@
-        public static Task RemoveFromGroup(
-            [HttpTrigger(AuthorizationLevel.Anonymous, "post")]HttpRequest req,
-            [SignalR(HubName = "simplechat")]IAsyncCollector<SignalRGroupAction> signalRGroupActions)
+        public Task OnExecutedAsync(FunctionExecutedContext executedContext, CancellationToken cancellationToken)
         {
-
-            var message = new JsonSerializer().Deserialize<ChatMessage>(new JsonTextReader(new StreamReader(req.Body)));
-
-            var decodedfConnectionId = GetBase64DecodedString(message.ConnectionId);
-
-            return signalRGroupActions.AddAsync(
-                new SignalRGroupAction
-                {
-                    ConnectionId = message.ConnectionId,
-                    UserId = message.Recipient,
-                    GroupName = message.Groupname,
-                    Action = GroupAction.Remove
-                });
+            return Task.CompletedTask;
         }
 
-        private static string GetBase64EncodedString(string source)
+        // gives customer a change to validate Function access token and add custom claims for ASRS access token.
+        // OnExecutingAsync runs before negotiate function but after arguments binding
+        public Task OnExecutingAsync(FunctionExecutingContext executingContext, CancellationToken cancellationToken)
         {
-            if (string.IsNullOrEmpty(source))
-            {
-                return source;
-            }
+            var serviceManager = StaticServiceHubContextStore.Get().ServiceManager;
+            var req = (HttpRequest)executingContext.Arguments["req"];
+            var connectionInfo = (SignalRConnectionInfo)executingContext.Arguments["connectionInfo"];
 
-            return Convert.ToBase64String(Encoding.UTF8.GetBytes(source));
-        }
+            // get header
+            var myHeader = req.Headers["myheader"];
 
-        private static string GetBase64DecodedString(string source)
-        {
-            if (string.IsNullOrEmpty(source))
+            // validate token
+            var result = TokenProvider.ValidateToken(req);
+            
+            if (result.Status == AccessTokenStatus.Valid)
             {
-                return source;
-            }
-
-            return Encoding.UTF8.GetString(Convert.FromBase64String(source));
-        }
-
+                // resolve the identity
+                string identity = result.Principal.Identity.Name;
 
-        public static class EventGridTriggerCSharp
-        {
-            [FunctionName("onConnection")]
-            public static Task EventGridTest([EventGridTrigger]EventGridEvent eventGridEvent,
-                [SignalR(HubName = "simplechat")]IAsyncCollector<SignalRMessage> signalRMessages)
+                // override connectionInfo argument
+                connectionInfo.AccessToken = serviceManager.GenerateClientAccessToken(Constants.HubName, identity, new List<Claim> {new Claim("myheader", myHeader)});
```

Comment:
todo: add a function in signalrconnectioninfo to update negotiate response instead of using servicemanager outside